xgcm/test/test_integrate.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import print_function
+import pytest
+
+from xgcm.grid import Grid
+from xgcm.test.datasets import datasets_grid_metric
+
+
+@pytest.mark.parametrize("gridtype", ["B", "C"])
+def test_integrate(gridtype):
+    ds_full = datasets_grid_metric()
+    ds = ds_full[gridtype]
+    grid = Grid(
+        ds,
+        coords=ds_full["coords"],
+        metrics=ds_full["metrics"],
+        # periodic=True,
+    )
+    # test tracer position
+    for axis, metric_name, dim in zip(
+        ["X", "Y", "Z", ["X", "Y"], ["X", "Y", "Z"]],
+        ["dx_t", "dy_t", "dz_t", "area_t", "volume_t"],
+        ["xt", "yt", "zt", ["xt", "yt"], ["xt", "yt", "zt"]],
+    ):
+        integrated = grid.integrate(ds.tracer, axis)
+        expected = (ds.tracer * ds[metric_name]).sum(dim)
+        assert integrated.equals(expected)
+        # test with tuple input if list is provided
+        if isinstance(axis, list):
+            integrated = grid.integrate(ds.tracer, tuple(axis))
+            expected = (ds.tracer * ds[metric_name]).sum(dim)
+            assert integrated.equals(expected)
+
+    # test u positon
+    if gridtype == "B":
+        for axis, metric_name, dim in zip(
+            ["X", "Y", ["X", "Y"]],
+            ["dx_ne", "dy_ne", "area_ne"],  # need more metrics?
+            ["xu", "yu", ["xu", "yu"]],
+        ):
+            integrated = grid.integrate(ds.u, axis)
+            expected = (ds.u * ds[metric_name]).sum(dim)
+            assert integrated.equals(expected)
+    elif gridtype == "C":
+        for axis, metric_name, dim in zip(
+            ["X", "Y", ["X", "Y"]],
+            ["dx_e", "dy_e", "area_e"],  # need more metrics?
+            ["xu", "yt", ["xu", "yt"]],
+        ):
+            integrated = grid.integrate(ds.u, axis)
+            expected = (ds.u * ds[metric_name]).sum(dim)
+            assert integrated.equals(expected)
+
+    # test v positon
+    if gridtype == "B":
+        for axis, metric_name, dim in zip(
+            ["X", "Y", ["X", "Y"]],
+            ["dx_ne", "dy_ne", "area_ne"],  # need more metrics?
+            ["xu", "yu", ["xu", "yu"]],
+        ):
+            integrated = grid.integrate(ds.v, axis)
+            expected = (ds.v * ds[metric_name]).sum(dim)
+            assert integrated.equals(expected)
+    elif gridtype == "C":
+        for axis, metric_name, dim in zip(
+            ["X", "Y", ["X", "Y"]],
+            ["dx_n", "dy_n", "area_n"],  # need more metrics?
+            ["xt", "yu", ["xt", "yu"]],
+        ):
+            integrated = grid.integrate(ds.v, axis)
+            expected = (ds.v * ds[metric_name]).sum(dim)
+            assert integrated.equals(expected)
```

Comment:
Sorry to be nit-picky here, but the style of these tests is important to me...

I would prefer to just have two separate tests, test_integrate_cgrid, test_integrate_bgrid, rather than this internal if / else logic.

Reply:
Roger that.

Reply:
I am implementing the changes but just out of curiosity? Is this so the pytest error messages are more specific? Just trying to learn along the way :-)

Reply:
I just don't like lots of if / else blocks in parameterized tests. Parameterization makes sense if you can run basically the same test on different inputs. But if you have to have big if / else blocks to deal with the different inputs, you don't gain anything by parameterizing other than unnecessary complexity. At that point you just have two separate functions already.

This is a stylistic choice, and there is no right and wrong here. But xgcm is my 👶and I just have these opinions.

Reply:
No thats fair. And it does make sense.